File: algorithms/twoSum/twoSum.cpp
// Source : https://oj.leetcode.com/problems/two-sum/
// Author : Hao Chen
// Date   : 2014-06-17

/********************************************************************************** 
* 
* Given an array of integers, find two numbers such that they add up to a specific target number.
* 
* The function twoSum should return indices of the two numbers such that they add up to the target, 
* where index1 must be less than index2. Please note that your returned answers (both index1 and index2) 
* are not zero-based.
* 
* You may assume that each input would have exactly one solution.
* 
* Input: numbers={2, 7, 11, 15}, target=9
* Output: index1=1, index2=2
* 
*               
**********************************************************************************/

class Solution {
public:
    /*
     *   The easy solution is O(n^2) run-time complexity.
     *   ```
     *       foreach(item1 in array) {
     *           foreach(item2 in array){
     *               if (item1 + item2 == target) {
     *                   return result
     *               }
     *           }
     *   ```
     *   
     *   We can see the nested loop just for searching, 
     *   So, we can use a hashmap to reduce the searching time complexity from O(n) to O(1)
     *   (the map's `key` is the number, the `value` is the position)
     *   
     *   But be careful, if there are duplication numbers in array, 
     *   how the map store the positions for all of same numbers?
     *
     */


    //
    // The implementation as below is bit tricky. but not difficult to understand
    //
    //  1) Traverse the array one by one
    //  2) just put the `target - num[i]`(not `num[i]`) into the map
    //     so, when we checking the next num[i], if we found it is exisited in the map.
    //     which means we found the second one.
    //      
    vector<int> twoSum(vector<int> &numbers, int target) {
        unordered_map<int, int> m;
        vector<int> result;
        for(int i=0; i<numbers.size(); i++){
            // not found the second one
            if (m.find(numbers[i])==m.end() ) { 
                // store the first one poisition into the second one's key
                m[target - numbers[i]] = i; 
            }else { 
                // found the second one
                result.push_back(m[numbers[i]]+1);
                result.push_back(i+1);
                break;
            }
        }
        return result;
    }
};

File: algorithms/python/Twosum.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-


# #Given an array of integers, return indices of the two numbers such that they add up to a specific target.
#
# #You may assume that each input would have exactly one solution, and you may not use the same element twice.
#
# #Example:
# #Given nums = [2, 7, 11, 15], target = 9,
#
# #Because nums[0] + nums[1] = 2 + 7 = 9,
# return [0, 1].
import math
import time


class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        list = []
        data = nums
        for i in nums:
            k = target - i
            if (k in nums) == True:
                print "K 值"
                print k, i;
                return
                #  list.append(target - i)

    def primeData(self, target):
        list = [];
        listPrimeData = []
        for i in range(3, target):
            for j in range(2, i):
                if i % j == 0:
                    list.append(i)
                    # print '%d ' % (i)
                    break;
            else:  # for函数也能做判断
                listPrimeData.append(i)

        return listPrimeData

    # expore more
    def primeDataOther(self, n):
        return filter(lambda x: not [x % i for i in range(2, int(math.sqrt(x)) + 1) if x % i == 0], range(2, n + 1))

    #
    # There are two sorted arrays nums1 and nums2 of size m and n respectively.
    #
    # Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
    #
    # Example 1:
    # nums1 = [1, 3]
    # nums2 = [2]
    #
    # The median is 2.0
    # Example 2:
    # nums1 = [1, 2]
    # nums2 = [3, 4]
    #
    # The median is (2 + 3)/2 = 2.5
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        x = len(nums1)
        y = len(nums2)
        z = x + y

        print z

        for i in range(len(nums1)):

    # twoSum(object,[1,2,3],5);
    listAlldata = range(100)
    print time.time()
    # print listAlldata
    print primeData(object, 10000)
    print time.time();
    # print primeData(object,10000)
    print primeDataOther(object, 10000)
    print time.time()
